File: Task1/src/cpp/rsa_dec.cpp
#include "rsa_utils.h"

#include <algorithm>
#include <cctype>
#include <exception>
#include <gmpxx.h>
#include <iostream>
#include <string>

namespace {
auto print_usage(const char *program) -> void {
	std::cerr << "Usage: " << program
			  << " <ciphertext_hex|\"Ciphertext: ciphertext_hex\"> "
				 "[private_key_file]\n";
}

auto trim_left(std::string value) -> std::string {
	value.erase(value.begin(),
				std::find_if(value.begin(), value.end(), [](unsigned char ch) {
					return !std::isspace(ch);
				}));
	return value;
}

auto normalize_ciphertext_arg(std::string value) -> std::string {
	const std::string prefix = "Ciphertext:";
	if (value.rfind(prefix, 0) == 0)
		value = trim_left(value.substr(prefix.size()));
	return value;
}
} // namespace

auto main(int argc, char *argv[]) -> int {
	try {
		if (argc < 2 || argc > 3) {
			print_usage(argv[0]);
			return 1;
		}

		std::string ciphertext_text = normalize_ciphertext_arg(argv[1]);
		const std::string private_key_file =
			argc >= 3 ? argv[2]
					  : textbookRSA::resolve_key_path("../rsa_private_key.txt");

		mpz_class ciphertext;
		if (ciphertext.set_str(ciphertext_text, 16) != 0)
			throw std::runtime_error("ciphertext must be a hexadecimal integer");

		mpz_class n, d;
		textbookRSA::read_private_key(n, d, private_key_file);
		if (ciphertext >= n)
			throw std::runtime_error("ciphertext is too large for this RSA key");

		const mpz_class message = textbookRSA::decrypt(ciphertext, d, n);
		std::cout << "Plaintext: " << textbookRSA::mpz_to_string(message)
				  << std::endl;
		return 0;
	} catch (const std::exception &ex) {
		std::cerr << "Error: " << ex.what() << std::endl;
		return 1;
	}
}

File: Task1/src/py/rsa_dec.py
#!/usr/bin/env python3
"""Textbook RSA 解密。

用法:  python rsa_dec.py <ciphertext_hex> [private_key_file]
       python rsa_dec.py "Ciphertext: <ciphertext_hex>" [private_key_file]

默认私钥文件为 Task1/RSA_Secret_Key.txt。打印 "Plaintext: ..."。
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import rsa_core

TASK1_DIR = Path(__file__).resolve().parents[2]


def normalize_ciphertext(value: str) -> str:
    prefix = "Ciphertext:"
    if value.startswith(prefix):
        value = value[len(prefix):].strip()
    return value


def main(argv: list[str]) -> int:
    if not 2 <= len(argv) <= 3:
        print("Usage: python rsa_dec.py <ciphertext_hex> [private_key_file]",
              file=sys.stderr)
        return 1

    ciphertext_text = normalize_ciphertext(argv[1])
    private_key_file = argv[2] if len(argv) >= 3 else TASK1_DIR / "RSA_Secret_Key.txt"

    try:
        ciphertext = int(ciphertext_text, 16)
    except ValueError:
        print("ciphertext must be a hexadecimal integer", file=sys.stderr)
        return 1

    n, d = rsa_core.read_private_key(private_key_file)
    if ciphertext >= n:
        print("ciphertext is too large for this RSA key", file=sys.stderr)
        return 1

    message = rsa_core.rsa_decrypt_int(ciphertext, d, n)
    plaintext = rsa_core.int_to_bytes(message).decode("utf-8", errors="replace")
    print("Plaintext:", plaintext)
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
